ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java

```diff
@@ -4696,4 +4696,37 @@ public void testMaterializedViewEnableRewriteNonBlocking() throws Exception {
     txnMgr.rollbackTxn();
     driver.run("drop materialized view mv_tab_acid");
   }
+
+  @Test
+  public void testMaterializedViewRebuildLockForSameMV() throws Exception {
```

Comment:
It seems that test case covers the problem. 
Some minor thoughts about it: 
- I would remove the unnecessary string concatenations, , like in line 4706 and 4710. I assume they are leftovers and they were multiple line statements at a previous iteration. 
- As the test case checks the proper locking, I would check the locks as well, like in `testMaterializedViewRebuildNoLocks`

Reply:
- Regarding the string concatenations, its just copied from above testcase but during code format they came in a single line. I believe they used string concatenation for readability purpose as the string is big.
- Here we are already checking the entries from MATERIALIZATION_REBUILD_LOCKS table which is related to actual problem.